### PR TITLE
Initial controller/raft integration

### PR DESCRIPTION
**Pull request by hachikuji**
A few highlights:

- We have a light shim for now to convert the raft layer to `MetaLogManager` (see `MetaLogRaftShim.java`)
- Now meta.properties includes both the broker.id and controller.id
- For now I have made the raft integration optional in `Kip500Controller` so that `KafkaClusterTestKit` does not depend on it
- `KafkaServer` startup had to be refactored significantly so that `RaftManager` could be started first in the lifecycle

Note that this patch also includes https://github.com/apache/kafka/pull/9732. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 1]
```diff
@@ -771,7 +771,7 @@ private void processDisconnection(List<ClientResponse> responses,
                     nodeId, disconnectState.remoteAddress());
                 break;
             case NOT_CONNECTED:
-                log.warn("Connection to node {} ({}) could not be established. Broker may not be available.", nodeId, disconnectState.remoteAddress());
+                log.debug("Connection to node {} ({}) could not be established. Broker may not be available.", nodeId, disconnectState.remoteAddress());
```
Comment (cmccabe) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:774]: what's the rationale for this change?

File: core/src/test/java/kafka/testkit/KafkaClusterTestKit.java [hunk 2]
```diff
@@ -17,8 +17,21 @@
...
+import kafka.server.Kip500Controller;
+import kafka.server.MetaProperties;
+import kafka.tools.StorageTool;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.common.utils.ThreadUtils;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.metalog.LocalLogManager;
```
Comment (rondagostino) [inline core/src/test/java/kafka/testkit/KafkaClusterTestKit.java:31]: Checkstyle violation: Disallowed import - org.apache.kafka.metalog.LocalLogManager.

**Review by rondagostino — COMMENTED** [review]
There are lots of failed tests due to a BrokerToControllerChannelManager issue recently fixed on the kip-500 branch.  Merging will fix these issues.

LocalLogManager seems to be broken, and this causes several failed tests.  For example:

LocalLogManagerTest (3 of 6 tests fail)
QuorumControllerTest (all 3 tests fail)
Kip500ControllerTest.testCreateControllersAndSendMetadataRequest
BrokerMetadataListenerTest.testLocalLogManagerEventQueue

Also commented on a few StorageToolTest failures.

File: core/src/main/scala/kafka/tools/StorageTool.scala [hunk 3]
```diff
@@ -198,9 +192,41 @@ object StorageTool extends Logging {
...
+    val brokerId = if (processRoles.contains(BrokerRole)) {
+      require(config.brokerId >= 0,
+        s"The `broker` process role requires non-negative `broker.id`, but found ${config.brokerId}")
+      Some(config.brokerId)
+    } else {
+      None
+    }
+
+    val controllerId = if (processRoles.contains(ControllerRole)) {
+      require(config.brokerId >= 0,
```
Comment (rondagostino) [inline core/src/main/scala/kafka/tools/StorageTool.scala:217]: `require(config.controllerId >= 0,`

It doesn't fix any of the 3 failing tests (indicating a gap in testing -- understandable at this point) but does fix this error if you try to format with a controller-only config file:

`bin/kafka-storage.sh format -c config/kip500-controller-node.properties --cluster-id 6bd37820-6745-4790-ae98-620300e1f61b`
```
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: The `controller` process role requires non-negative `controller.id`, but found 3000
	at kafka.tools.StorageTool$.buildMetaProps(StorageTool.scala:218)
```

**Comment by rondagostino** [comment]
Latest config files are now reamed on the kip-500 branch via https://github.com/confluentinc/kafka/pull/470, so this PR will need to merge those changes and then convert `controller.connect` to `controller.quorum.voters` in the 3 new config files.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java (hunk 1 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:774]: I'll revert it. It was just really spammy when I was testing failover.

**Comment by rondagostino** [comment]
Setting the follwinfg definition in `BrokerLifecycleManager.scala` will allow this branch to successfully start a controller and a broker:

```
  private def handleBrokerHeartbeatResponse(response: BrokerHeartbeatResponse): Option[Errors] = {
    if (response.data().errorCode() != 0) {
      val errorMsg = Errors.forCode(response.data().errorCode())
      error(s"Broker heartbeat failure: $errorMsg")
      Some(errorMsg)
    } else {
      val origState = currentState
      currentState = jmetadata.BrokerState.fromValue(response.data().nextState())
      // notifying brokerMetadataListener when fencing/unfencing is what was missing
      if (origState == jmetadata.BrokerState.FENCED && currentState == jmetadata.BrokerState.RUNNING) {
        brokerMetadataListener.put(FenceBrokerEvent(brokerEpoch(), false))
      } else if (origState == jmetadata.BrokerState.RUNNING && currentState == jmetadata.BrokerState.FENCED) {
        brokerMetadataListener.put(FenceBrokerEvent(brokerEpoch(), true))
      }
      _lastSuccessfulHeartbeatTime = time.nanoseconds
      None
    }
  }
```

This doesn't help with the merge of the kip-500 branch, of course, because that branch has eliminated the concept of fencing altogether.  But the kip-500 branch has other problems (notably the broker can't identify the controller node, so the registration request never makes it to the controller).  I suspect if we perform the merge we will have something here that works.

**Comment by cmccabe** [comment]
> But the kip-500 branch has other problems (notably the broker can't identify the controller node, so the registration request never makes it to the controller).

Hmm, this is working for me.  What issue did you see?

**Comment by rondagostino** [comment]
> Hmm, this is working for me. What issue did you see?

@cmccabe For the kip-500 branch, the broker never identifies the remote controller as node 3000 when I start a separate broker and controller -- the broker always has `-1` as the controller's  node ID.  The broker does identify the controller node as being `3000` if I start a **combined** broker and controller on the kip-500 branch, but then the registration request times out after 30 seconds, and after the 3rd retry times out (2 minutes after the initial registration attempt was sent) the process shuts down.

On this branch, I see the same behavior with a combined broker+controller: the broker identifies node 3000 as the controller and sends 4 registration requests across 2 minutes that all timeout and then the process shuts down.    However, this branch works with separate broker and controller: registration and heartbeat requests succeed.

**Comment by rondagostino** [comment]
> this branch works

Correction, this branch with appropriate fixes and the `kip-500` branch merged in as per https://github.com/confluentinc/kafka/pull/472.  We might see the same behavior on this PR with fixes applied and without the kip-500 merge but haven't gone back to confirm that.

**Comment by hachikuji** [comment]
Closing this since #470 was merged.
